content/generators/sudoku.md

```diff
@@ -0,0 +1,163 @@
+
+    numbers  = shuffle(range(1,rank*rank+1))
+
+    board = [ [numbers[pattern(row,col)] for col in cols] for row in rows ]
+    board = np.asarray(board)
+    return board
+
+# Now we want to create the structure of a sudoku graph which we can fill up with the random grid we generated.
+# Note how we store the three different kinds of edges separately, this will be useful for visualization
+def empty_sudoku_graph(n):
+    blank_board = np.arange(n*n*n*n).reshape(n*n,n*n)
+    def get_edges(nodes):
+        edges = [(a, b) for idx, a in enumerate(nodes) for b in nodes[idx + 1:]]
+        return edges
+    box_edges=[]
+    row_edges=[]
+    column_edges=[]
+    for iteration in range(n*n):
+        i = iteration//n
+        j = iteration%n
+        column = np.asarray(blank_board)[:,iteration]
+        row = np.asarray(blank_board)[iteration,:]
+        box=blank_board[(i)*n:(i+1)*n,(j)*n:(j+1)*n].flatten()
+        box_edges += get_edges(box)
+        row_edges += get_edges(row)
+        column_edges += get_edges(column)
+    return box_edges, row_edges, column_edges
+
+# We have the edges and the nodes, now we want a networkx graph
+def draw_networkx_graph(n, board, box_edges, row_edges, column_edges, layout='grid'):
+    G = nx.empty_graph(n*n*n*n)
+    if layout=='circular':
+        pos=nx.circular_layout(G)
+    if layout=='spring':
+        pos=nx.spring_layout(G)
+    if layout=='grid':
+        pos = dict(zip(list(G.nodes()), nx.grid_2d_graph(n*n,n*n)))
+
+    mapping = dict(zip(G.nodes(), board.flatten()))
+    plt.figure(1,figsize=(12,12))
+    nx.draw(G, pos, labels=mapping, with_labels=True, node_color='white')
+    nx.draw_networkx_edges(G,pos,edgelist=box_edges,width=2, alpha=0.5, edge_color="tab:red")
+    nx.draw_networkx_edges(G,pos,edgelist=row_edges,width=2, alpha=0.5, edge_color="tab:green")
+    nx.draw_networkx_edges(G,pos,edgelist=column_edges,width=2, alpha=0.5, edge_color="tab:blue")
+    plt.show()
+
+```
+
+
+```{code-cell} ipython3
+def sudoku(n, layout):
+    board = generate_random_sudoku(n)
+    box_edges, row_edges, column_edges = empty_sudoku_graph(n)
+    draw_networkx_graph(n, board, box_edges, row_edges, column_edges, layout)
+```
+
+in this view we can see th three different edge colors as the three different ways two nodes can be adjacent in a sudoku graph. Observe that no edge connects the same two numbers, this is what it means to satisfy the constraints of a sudoku problem
+
+
+```{code-cell} ipython3
+sudoku(3, 'grid')
+```
+
+We couldn't see a lot of edges in the first figure because they fell on top of each other, this is perhaps a better view, if you want to look at this in an animated way, check [this old reddit post](https://reddit.com/r/dataisbeautiful?utm_content=subreddit&utm_medium=post_embed&utm_name=8775be457c4149bd9d15290fa5723658&utm_source=embedly&utm_term=6ty4vf) although the code isn't available there and it's an old link. This is a simpler way to do it in networkx
```

Comment:
I'm -1 on linking to the old reddit post. While the visualization is nice, it doesn't have a lot of pedagogical value. My vote would be to remove it for now, or better yet - try a tool like [bokeh](https://bokeh.org/) to implement the interactive visualization. If you were interested in doing so, there's [an example in the bokeh gallery that might prove useful](https://docs.bokeh.org/en/latest/docs/gallery/network_graph.html)